### DISCUSSION TITLE
Allow type narrowing via any when iterating over immutable types?

### DISCUSSION THREAD

**Discussion by ringohoffman** [microsoft/pyright#4832] [Ideas]
I have a few variables I want to assert are not `None` without having to type out a giant conditional e.g. `a is None or b is None ...`. I had the thought to use `any` to simplify this for me:

```python
from typing import Any, Optional


def f() -> Any:
    return 1


a: Optional[int] = f()
b: Optional[int] = f()

if any(i is None for i in (a, b)):
    raise ValueError("should all be not None")

reveal_type(a)  # Type of "a" is "int | None"
reveal_type(b)  # Type of "b" is "int | None"
```

but the types didn't get narrowed through the `any`. Would it be possible to support this?

**Comment by erictraut** [comment]
This isn't one of the supported patterns supported by pyright. For a full list of supported forms, see [this documentation](https://microsoft.github.io/pyright/#/type-concepts-advanced?id=type-guards). Pyright already supports more type narrowing forms than any other Python type checker, at least to my knowledge.

We do occasionally add support for new patterns, but each one requires very specific logic and often slows down overall type analysis, so we do so very sparingly. Supporting the specific pattern `any (i is None for i in (<tuple>))` would need to be incredibly specific to that form, and this particular form is quite unusual. A more idiomatic form might be `if None in (a, b)`.

The proposed pattern also involves a function call rather than operators that are built into the language, so it would involve hard-coding knowledge of the function call — something we try to avoid.

You can leverage one of the existing narrowing forms to accomplish what you want:
```python
    if a is None or b is None:
        raise ValueError("should all be not None")
```

Or you can build your own user-defined type guards using PEP 647:
```python
def no_nones(a: tuple[T | None, ...]) -> TypeGuard[tuple[T, ...]]:
    return None not in a


def func(a: int | None, b: int | None):
    x = (a, b)
    if not no_nones(x):
        raise ValueError("should all be not None")

    reveal_type(x[0])
```

**Reply by ringohoffman** [reply]
I really like `None in (a, b)`! It's a shame it doesn't narrow types. Thanks for the explanation and documentation links.
